### PR TITLE
Example go server

### PR DESCRIPTION
**Pull request by flumonion**

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -45,4 +45,39 @@ composer install
 ## General
 
 - PHP version 8.1
-- Use strict types `declare(strict_types=1);`
\ No newline at end of file
+- Use strict types `declare(strict_types=1);`
+
+***
+
+## Example Ecommerce store
```
Comment (flumonion) [inline readme.md]: We need to include the built binaries and instructions to run these binaries.
We should not expect the developer to know about Go or even have it installed.

You can keep the below instructions, perhaps lets move them to the example store directory...

**Comment by mozey** [comment]
@Ayano2000 I see you've chosen to use [gorilla/mux](https://github.com/gorilla/mux), that's fine. However, elsewhere ([stock2shop/api](https://github.com/stock2shop/api), [apifact_server](https://github.com/stock2shop/apifact_server), and [connector_go](https://github.com/stock2shop/connector_go)) we're using [julienschmidt/httprouter](https://github.com/julienschmidt/httprouter), see [simple usage examples here](https://github.com/mozey/httprouter-example). 

Wouldn't it make sense to use the same lib here as well?

Not that relevant for this discussion, but worth nothing. Avoid using [gorilla compress handler middleware](https://github.com/gorilla/handlers/blob/master/compress.go) for gzipping, I've had issues with that one in the past. If I recall it sometimes incorrectly served the HTTP response as a file, instead of text, but I can't remember the exact use case. [NYTimes/gziphandler](https://github.com/NYTimes/gziphandler) is a better option, and also used in the repos mentioned above

File: readme.md (hunk 1 above)
Comment (mozey) [reply readme.md]: > We need to include the built binaries

Keeping the binaries in the repo is not a good idea. If you have to make the binaries available, you can 
- link to a public S3 bucket with the binaries
- upload the binaries when [tagging a release](https://github.com/stock2shop/connector/releases)

**Comment by flumonion** [comment]
@mozey those are good suggestions.

It was my idea to include the binaries, so it is really easy for the developer to get setup.
I did not want Go as a dependency.

Hosting them is fine, question is where?
Considering ~15mb of data for the binaries, should we not just leave them here?

**Comment by mozey** [comment]
> should we not just leave them here?

You could, but it's a bit weird... and not best practice. Over time it can bloat the repo. I've already made [two suggestions here](https://github.com/stock2shop/connector/pull/6#discussion_r999202962). See [managing releases in a repository](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository): *"You can create new releases with release notes, @mentions of contributors, and links to binary files, as well as edit or delete existing releases."*

For example, previously I moved binaries from stock2shop/apifact to the [apifact-bin bucket](https://github.com/stock2shop/apifact#apifact-bin)

**EDIT** I notice above bucket is still empty. These binaries currently exist at `apifact.com:/home/ubuntu/bin`, I'll make sure they're backed up to the bucket as per the issue linked in the docs

**Comment by mozey** [comment]
> This bucket must not be publicly available

For the connector you probably want to use a public bucket. The github release can then link to it

**Comment by flumonion** [comment]
@Ayano2000 I will create a public bucket.
For the time being, let's continue with testing etc. 
When we ready for V1.0 we can tag it with links and put it in readme and remove binaries.
We probably need to do [something like this](https://stackoverflow.com/questions/28312439/git-gc-aggressive-prune-all-does-not-remove-big-file-from-repository) to reduce the size. Need to research it a bit more.

**Comment by mozey** [comment]
> We probably need to do [something like this](https://stackoverflow.com/questions/28312439/git-gc-aggressive-prune-all-does-not-remove-big-file-from-repository)

I'm not sure about those steps... As per [this comment](https://github.com/stock2shop/apifact/issues/90#issuecomment-1283913916), see [notes about removing history here](https://www.mozey.co/post/git/#removing-history). I've done those steps before, but obviously on your own risk when doing stuff like that. Make sure you have a backup first

**EDIT** Also see [comment about checksums](https://github.com/stock2shop/apifact/issues/90#issuecomment-1283852511)

File: src/ChannelProducts.php [hunk 2]
```diff
@@ -12,40 +12,65 @@ public function sync(
...
-                    $i->channel_image_code = (string)$i->id;
-                }
-                file_put_contents($path, json_encode($product));
+                $touchProducts[] = Helper::createPayloadProduct($product);
+            }
+        }
+
+        // create/update products
+        if (count($touchProducts) > 0) {
+            $ch = curl_init();
```
Comment (flumonion) [inline src/ChannelProducts.php]: use guzzle, include it as dependancy.

File: src/ChannelProducts.php [hunk 3]
```diff
@@ -12,40 +12,65 @@ public function sync(
...
+            curl_setopt($ch, CURLOPT_URL, 'http://localhost:1234/products');
```
Comment (flumonion) [inline src/ChannelProducts.php]: The host should be configurable using channel meta

File: src/ChannelProducts.php [hunk 4]
```diff
@@ -54,29 +79,50 @@ public function get(
...
-                }
-                $data['channel']['success'] = true;
-                foreach ($data['variants'] as $k => $v) {
-                    $data['variants'][$k]['channel']['success'] = true;
-                }
-                foreach ($data['images'] as $k => $v) {
-                    $data['images'][$k]['channel']['success'] = true;
-                }
-                $results['channel_products'][] = $data;
+        foreach ($data as $p) {
```
Comment (flumonion) [inline src/ChannelProducts.php]: create transform class

File: src/ChannelProducts.php [hunk 5]
```diff
@@ -85,35 +131,49 @@ public function getByCode(
...
+        $ch = curl_init();
+        curl_setopt($ch, CURLOPT_URL, 'http://localhost:1234/products');
+        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "GET");
+        curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($codes));
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+        $server_output = curl_exec($ch);
+        curl_close($ch);
+        $data = json_decode($server_output);
+
+        foreach ($channelProducts->channel_products as $value => $product) {
```
Comment (flumonion) [inline src/ChannelProducts.php]: transform class?

File: readme.md [hunk 6]
```diff
@@ -37,12 +37,52 @@ cd share
 composer install
 ```
 
-## running Tests
+## running tests
+
+First start the example server provided, binaries for `macOS`, `windows` and `linux` systems can be found at `connector/example_ecommerce_store/bin/`.
```
Comment (flumonion) [inline readme.md]: write out commands

File: composer.json [hunk 7]
```diff
@@ -1,39 +1,40 @@
...
+  "require": {
+    "php": ">=8.1",
+    "stock2shop/share": "dev-master",
+    "vlucas/phpdotenv": "^5.4",
+    "monolog/monolog": "^2.8.0",
+    "maxbanton/cwh": "^2.0.2",
+    "aws/aws-sdk-php": "^3.238"
+  },
+  "require-dev": {
+    "guzzlehttp/guzzle": "^7.0",
```
Comment (flumonion) [inline composer.json]: guzzle is not a dev requirement?

File: src/ChannelProducts.php [hunk 8]
```diff
@@ -4,48 +4,62 @@
...
     public function sync(
         Share\DTO\ChannelProducts $channelProducts,
         Share\DTO\Channel         $channel
     ): Share\DTO\ChannelProducts {
-        Helper::setDataDir();
-        foreach ($channelProducts->channel_products as $product) {
-            $path = Helper::getProductPath($product);
+        $url = self::getServerUrlFromChannelMeta($channel->meta);
+        if ($url == "") {
+            return new Share\DTO\ChannelProducts([]);
```
Comment (flumonion) [inline src/ChannelProducts.php]: We need to return all products but marked as unsuccessful.

File: src/DemoAPI/API.php [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    /**
+     * @return Product[]
+     * @throws GuzzleException
+     */
+    public function getProducts(string $fromID, int $limit): array
+    {
+        // create guzzle client
+        $client = new Client([
```
Comment (flumonion) [inline src/DemoAPI/API.php]: guzzle client set in constructor rather?
The methods are not static so he constructor will always be called/

**Comment by mozey** [comment]
Know this has been merged, just occurred to me I created a ["Mock Server" issue](https://github.com/stock2shop/cart/issues/4) w.r.t. the proposed stock2shop cart. Kind of related to what was done here so linking for future reference

**Pull request merged**